### PR TITLE
ENH: Jones and Faddy Skew-T distribution

### LINKED ISSUE #18947: ENH: Jones and Faddy Skew-T distribution
### Is your feature request related to a problem? Please describe.

Jones and Faddy introduced a skew extension of the t distribution, which includes the t distribution as a special case. Their main paper discussing it is available [here](https://academic.oup.com/jrsssb/article/65/1/159/7110653?login=false).

The Jones and Faddy skew-t distribution can provide a good fit for data which is heavy tailed and skewed such as asset returns in finance. This makes it suited for Monte Carlo simulations in order to calculate the market risk associated with financial assets. I've used it for this purpose both in personal projects and at work (using implementations built independent of SciPy, and also recently implementing it myself using SciPy machinery).

It's also a distribution which has been applied more widely: their paper on the distribution has been cited 254 times. Jones and Faddy provide in their paper two applications where the skew-t is fit to: (1) the strength of glass fibres and (2) blood flow data.

It would be useful to have an implementation as part of SciPy so that it's easier for others to make use of.

### Describe the solution you'd like.

An implementation in SciPy of the Jones and Faddy Skew-T distribution. It's quite a tractable distribution, with analytical expressions for the mode, pdf, cdf, inverse cdf, and moments (all derived and presented in Jones and Faddy's 2003 paper). The paper also gives an expression for a skew-t random variable in terms of a beta random variable, which is useful for simulating samples from a skew-t distribution.

It would be good for the implementation to make use of all the above, to make computations more efficient.

I would propose calling the implementation a `jf_skew_t` distribution, or Jones and Faddy Skew-T distribution. This distinguishes it from other skew-t extensions which exist, for example the two discussed in Jones and Faddy's [paper](https://academic.oup.com/jrsssb/article/65/1/159/7110653?login=false) in sections 5.1 and 5.2.

### Describe alternatives you've considered.

_No response_

### Additional context (e.g. screenshots, GIFs)

_No response_

### PR DESCRIPTION
**Pull request by JWTownhill**
#### Reference issue
Closes https://github.com/scipy/scipy/issues/18947

#### What does this implement/fix?
This PR implements Jones and Faddy's skewed extension of the t distribution. This distribution is useful when describing data which is both heavy tailed and skewed.

### CODE REVIEW

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:5384]: In #17624 , I noticed that using the RNG's [beta](https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.beta.html#numpy.random.Generator.beta) function is faster than calling scipy's `beta` distribution. Could you test it here as well?

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:5394]: For better performance, I prefer to call the scipy `special` functions directly. Here we should try `sc.betainc`

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3712]:
```suggestion
```

These are already covered in the generic tests: https://github.com/scipy/scipy/blob/main/scipy/stats/tests/test_continuous_basic.py#L580

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3704]: How were the expected values computed?

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
Thanks for the PR @JWTownhill !

Typically, new distributions have to be proposed on the mailing list first. The distribution seems to me important enough though due to the number of citations and the PR is already there, so I had a first look.

The most important thing is probably the name of the distribution. I would opt for `skewt`, if there are no other popupar skewed generalizations of the central t distribution.

Another point to consider: could you add a `_logpdf` function?

File: scipy/stats/tests/test_distributions.py
Comment (JWTownhill) [reply scipy/stats/tests/test_distributions.py:3704]: These were computed by me in a Jupyter notebook. I could also have used some parameters which give simpler stats, e.g. `a=b=3` would give stats `[0, 1.5, 0, 3]` (which is easy to verify because a=b=3 gives a t distribution with 6 degrees of freedom).

It seemed like a good idea to have this test mainly to flag any issues that changes to the code might cause.

I'll take others' lead on whether it's needed or not though.

**Comment by JWTownhill** [comment]
> Thanks for the PR @JWTownhill !

:) Thanks for your comments and suggested changes.

> Typically, new distributions have to be proposed on the mailing list first. The distribution seems to me important enough though due to the number of citations and the PR is already there, so I had a first look.

Aha, thank you, I have now tried to send something around on the mailing list. I wasn't aware it existed before - I had been following this guide in the first instance: https://docs.scipy.org/doc/scipy/dev/contributor/adding_new.html#adding-a-new-statistics-distribution

> The most important thing is probably the name of the distribution. I would opt for `skewt`, if there are no other popupar skewed generalizations of the central t distribution.

I'm not confident this is the most popular skewed generalised t distribution, but I find it easier to use than [this](https://academic.oup.com/jrsssb/article/65/2/367/7092856) multivariate version by Azzalini which has 1.1k citations for the paper. In the Jones and Faddy 2003 paper they discuss a couple of other skewed t distributions which exist (including the Azzalini one).

It seemed safer to opt for `jf_skew_t` rather than calling the Jones and Faddy version simply `skewt` which implies it is the 'default' version.

> Another point to consider: could you add a `_logpdf` function?

Perhaps I could! What is the advantage of it though? Do you think a better job can be done than the default behaviour of taking the log of the pdf? It's not obvious to me.

Thanks again for your comments, I will look through your suggested changes.

File: scipy/stats/tests/test_distributions.py
Comment (JWTownhill) [reply scipy/stats/tests/test_distributions.py:3712]: Thanks, I'll remove. I got the idea from the test here: https://github.com/scipy/scipy/blob/b5e02b70c9b49d8b6283542be8786f0cdd734f2a/scipy/stats/tests/test_distributions.py#L6858-L6862 

Perhaps there was a specific reason they wanted to test more quantiles? I don't think it's important for this distribution though, so I'll remove.

**Comment by dschmitz89** [comment]
> > Thanks for the PR @JWTownhill !
> 
> :) Thanks for your comments and suggested changes.
> 
> > Typically, new distributions have to be proposed on the mailing list first. The distribution seems to me important enough though due to the number of citations and the PR is already there, so I had a first look.
> 
> Aha, thank you, I have now tried to send something around on the mailing list. I wasn't aware it existed before - I had been following this guide in the first instance: https://docs.scipy.org/doc/scipy/dev/contributor/adding_new.html#adding-a-new-statistics-distribution
> 
> > The most important thing is probably the name of the distribution. I would opt for `skewt`, if there are no other popupar skewed generalizations of the central t distribution.
> 
> I'm not confident this is the most popular skewed generalised t distribution, but I find it easier to use than [this](https://academic.oup.com/jrsssb/article/65/2/367/7092856) multivariate version by Azzalini which has 1.1k citations for the paper. In the Jones and Faddy 2003 paper they discuss a couple of other skewed t distributions which exist (including the Azzalini one).
> 
> It seemed safer to opt for `jf_skew_t` rather than calling the Jones and Faddy version simply `skewt` which implies it is the 'default' version.
> 

Ok, that sounds convincing to me.

> > Another point to consider: could you add a `_logpdf` function?
> 
> Perhaps I could! What is the advantage of it though? Do you think a better job can be done than the default behaviour of taking the log of the pdf? It's not obvious to me.
> 

Taking the logarithm of the beta function avoids intermediate overflow:

```python
from mpmath import mp
import numpy as np
from scipy import special as sc

def _pdf(x, a, b):
    c = 2 ** (a + b - 1) * sc.beta(a, b) * np.sqrt(a + b)
    d1 = (1 + x / np.sqrt(a + b + x ** 2)) ** (a + 0.5)
    d2 = (1 - x / np.sqrt(a + b + x ** 2)) ** (b + 0.5)
    return d1 * d2 / c

def _logpdf(x, a, b):
    c = np.log(2) * (a + b - 1) + sc.betaln(a, b) + 0.5 * np.log(a + b)
    d1 = np.log1p(x / np.sqrt(a + b + x ** 2)) * (a + 0.5)
    d2 = np.log1p(-x / np.sqrt(a + b + x ** 2)) * (b + 0.5)
    return d1 + d2 - c

def pdf_mpmath(x, a, b):
    x = mp.mpf(x)
    a = mp.mpf(a)
    b = mp.mpf(b)

    c = 2 ** (a + b - mp.one) * mp.beta(a, b) * mp.sqrt(a + b)
    d1 = (mp.one + x / np.sqrt(a + b + x ** 2)) ** (a + 0.5)
    d2 = (mp.one - x / np.sqrt(a + b + x ** 2)) ** (b + 0.5)
    return float(d1 * d2 / c)

mp.dps = 500

x, a, b = 10, 1e3, 2e3

np.exp(_logpdf(x, a, b)), pdf_mpmath(x, a, b) # this works
>> (3.667440225059364e-175, 3.667440225071735e-175)

_pdf(x, a, b) # this fails with
      1 def _pdf(x, a, b):
----> 2     c = 2 ** (a + b - 1) * sc.beta(a, b) * np.sqrt(a + b)
      3     d1 = (1 + x / np.sqrt(a + b + x ** 2)) ** (a + 0.5)
      4     d2 = (1 - x / np.sqrt(a + b + x ** 2)) ** (b + 0.5)

OverflowError: (34, 'Result too large')
``` 

Admittedly, these are extreme parameter values but at some point there will be users testing out extreme parameter ranges. Often, doing intermediate calculations in logspace avoids such problems in statistical routines. If possible, let's try to avoid such issues from the beginning.

This is also another PR where #17697 would come in handy for the survival and inverse survival functions but that one should not hold this up.

**Comment by steppi** [comment]
> Another point to consider: could you add a `_logpdf` function?

I think it's fine to leave `_logpdf` for a follow-up. Technically a more accurate way to handle the intermediate overflow here would be to use the [Lanczos approximation](https://www.boost.org/doc/libs/1_82_0/libs/math/doc/html/math_toolkit/lanczos.html) which is available in `special` as a private function. This could be left to the far future though if we ever decide we need that level of accuracy.

**Review by steppi — CHANGES_REQUESTED** [review]
This looks like it’s in good shape. The only major suggestion is to add a test comparing to a reference implementation. There’s an R implementation you can use; see the comment. Adding `_logpdf` can be saved for a follow up PR imo.

File: scipy/stats/tests/test_distributions.py
Comment (JWTownhill) [reply scipy/stats/tests/test_distributions.py:3698]: Thank you for the example. I've now added a comparison with the pdf (`dST5`) values from the R implementation.

Interestingly, the GAMLSS implementation uses the 'alternative parameterisation' that Jones and Faddy provide in their paper - but I don't think that's canonical and it's easy to convert the parameters. So I think sticking with the 'a' and 'b' parameters is the correct choice for SciPy.

File: scipy/stats/tests/test_distributions.py
Comment (JWTownhill) [reply scipy/stats/tests/test_distributions.py]: I've removed this test now.

**Comment by JWTownhill** [comment]
Please could someone let me know the correct workflow to merge the PR once it gets approved? Most of the time I would squash my commits so they become just one commit and rebase my branch before merging it, but I'm not sure how things are usually done for SciPy. Thanks

**Comment by dschmitz89** [comment]
> Please could someone let me know the correct workflow to merge the PR once it gets approved? Most of the time I would squash my commits so they become just one commit and rebase my branch before merging it, but I'm not sure how things are usually done for SciPy. Thanks

This will be taken care of by a maintainer, no need to worry about that on your end.

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:3740]:
```suggestion
        assert_allclose(pdf, stats.jf_skew_t(a, b).pdf(x), rtol=1e-14)
```

[`assert_allcose's`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html) default tolerance is `1e-7`. If possible, we try to get lower than that. Could you test out how low we can go here?

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
This looks almost done to me, I only have one remaining comment about the test tolerance.

File: scipy/stats/tests/test_distributions.py
Comment (JWTownhill) [reply scipy/stats/tests/test_distributions.py:3740]: Sure I'll have a go. I just tried with 1e-14 as suggested, but that didn't work so I'll play around with a few powers of 10 to see.

Comment (JWTownhill) [reply scipy/stats/tests/test_distributions.py:3740]: We can push it quite far - `1e-12` succeeds but `1e-13` fails, so I'll select `rtol=1e-12` :)

**Review by steppi — CHANGES_REQUESTED** [review]
@JWTownhill, sorry I'd missed something earlier. You should add a `_shape_info` method to `jf_skew_t_gen`. This makes the class aware of the names, types and supports of the shape parameters and is needed for fitting to work properly.

In your case `_shape_info` will be the same as that for [beta_gen](https://github.com/scipy/scipy/blob/baf7f2e3e53edd60db43d40a404bf1f74bf2621a/scipy/stats/_continuous_distns.py#L669-L672). See [here](https://github.com/scipy/scipy/blob/baf7f2e3e53edd60db43d40a404bf1f74bf2621a/scipy/stats/_distn_infrastructure.py#L1559-L1561) for the meaning of the `_ShapeInfo` params in case you're curious.

~It would also be good to add an `_argcheck` method which validates that `a` and `b` are both greater than zero. Take a look at some of the other `_argcheck` methods.~ Nevermind, sorry, I forgot `_argcheck` defaults to checking positivity so you don't need to explicitly add this. You just need to add `_shape_info`.

**Comment by steppi** [comment]
I'm seeing a relevant test failure. There are a lot of tests failing for unrelated reasons that made this easier to miss.

```
FAILED stats/tests/test_fit.py::test_cont_fit[MM-jf_skew_t-arg48] - ValueError: Method of moments
encountered a non-finite distribution moment and cannot continue. Consider trying method='MLE'.
```

In `stats/tests/test_fit.py` you need to add `jf_skew_t` to `mm_failing_fits` here:

https://github.com/scipy/scipy/blob/65635dd55dca0962ab1e3260500b84306fc6500c/scipy/stats/tests/test_fit.py#L58-L68

after that everything should be good to go.

**Comment by JWTownhill** [comment]
> In `stats/tests/test_fit.py` you need to add `jf_skew_t` to `mm_failing_fits` here:

Thank you, I had missed this because I'd been running only the default 'fast' tests locally. Now added to the list.

**Review by steppi — APPROVED** [review]
Looks good @JWTownhill. Really nice work here!

**Comment by steppi** [comment]
Sorry, before this goes in someone needs to look over the docs as well. @dschmitz89, do you have bandwidth?

File: doc/source/tutorial/stats/continuous_jf_skew_t.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_jf_skew_t.rst:33]:
```suggestion
   :doi:`10.1111/1467-9868.00378`
```
Here and in the tutorial to make this render correctly.

File: doc/source/tutorial/stats/continuous_jf_skew_t.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_jf_skew_t.rst:15]: Looks like it matches:
image

File: doc/source/tutorial/stats/continuous_jf_skew_t.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_jf_skew_t.rst:16]: Seems to match
image

File: doc/source/tutorial/stats/continuous_jf_skew_t.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/continuous_jf_skew_t.rst:14]: Agrees with the API documentation.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:5351]: Matches
image

But maybe a bit too closely. I'd reword some of this slightly.

**Review by mdhaber — APPROVED** [review]
Documentation looks good with a few minor adjustments. (I don't need to review again.)

File: doc/source/tutorial/stats/continuous_jf_skew_t.rst
Comment (JWTownhill) [reply doc/source/tutorial/stats/continuous_jf_skew_t.rst:19]: The domain of `a` and `b` is included on lines 7 and 8

File: scipy/stats/_continuous_distns.py
Comment (JWTownhill) [reply scipy/stats/_continuous_distns.py:5351]: Ok, I've changed it to (in both the doc string and the tutorial file):
```
When :math:`a<b`, the distribution is negatively skewed, and when 
:math:`a>b`, the distribution is positively skewed. If :math:`a=b`, then 
we recover the `t` distribution with :math:`2a` degrees of freedom.
```

**Comment by steppi** [comment]
@JWTownhill, before I approve the workflows, can you push another commit which contains `[skip cirrus] [skip actions]` in the message. This will skip CI jobs which aren't relevant for a doc only change. You can do something like

```bash
git commit --allow-empty -m "EMPTY: [skip cirrus] [skip actions]"
```

**Pull request merged**
